src/data_source/sql.rs

```diff
@@ -608,134 +649,408 @@ impl VersionedDataSource for SqlDataSource {
         }
-        // If we're trying to roll back a transaction, something has already gone wrong and we're
-        // trying to recover. If we're unable to revert the changes and recover, all we can do is
-        // panic.
-        self.client
-            .batch_execute("ROLLBACK")
-            .await
-            .expect("DB rollback succeeds");
-        self.tx_in_progress = false;
+        self.leaf_stream.revert();
+        self.block_stream.revert();
     }
 }
 
 #[async_trait]
-impl<Types, I> AvailabilityDataSource<Types, I> for SqlDataSource
+impl<Types, I> AvailabilityDataSource<Types, I> for SqlDataSource<Types, I>
 where
     Types: NodeType,
     I: NodeImplementation<Types>,
     Block<Types>: QueryableBlock,
 {
-    type LeafStream = BoxStream<'static, LeafQueryData<Types, I>>;
-    type BlockStream = BoxStream<'static, BlockQueryData<Types>>;
+    type LeafStream = BoxStream<'static, QueryResult<LeafQueryData<Types, I>>>;
+    type BlockStream = BoxStream<'static, QueryResult<BlockQueryData<Types>>>;
 
-    type LeafRange<'a, R> = BoxStream<'a,  QueryResult<LeafQueryData<Types, I>>>
+    type LeafRange<'a, R> = BoxStream<'static, QueryResult<LeafQueryData<Types, I>>>
     where
         Self: 'a,
         R: RangeBounds<usize> + Send;
-    type BlockRange<'a, R>= BoxStream<'a,  QueryResult<BlockQueryData<Types>>>
+    type BlockRange<'a, R>= BoxStream<'static, QueryResult<BlockQueryData<Types>>>
     where
         Self: 'a,
         R: RangeBounds<usize> + Send;
 
-    async fn get_leaf<ID>(&self, _id: ID) -> QueryResult<LeafQueryData<Types, I>>
+    async fn get_leaf<ID>(&self, id: ID) -> QueryResult<LeafQueryData<Types, I>>
     where
         ID: Into<LeafId<Types, I>> + Send + Sync,
     {
-        todo!()
+        let (where_clause, param): (&str, Box<dyn ToSql + Send + Sync>) = match id.into() {
+            ResourceId::Number(n) => ("height = $1", Box::new(n as i64)),
+            ResourceId::Hash(h) => ("hash = $1", Box::new(h.to_string())),
+        };
+        let query = format!("SELECT leaf, qc FROM leaf WHERE {where_clause}");
+        let row = self
+            .client
+            .query_opt(&query, &[&*param])
+            .await
+            .map_err(|err| QueryError::Error {
+                message: err.to_string(),
+            })?
+            .context(NotFoundSnafu)?;
+        parse_leaf(row)
     }
-    async fn get_block<ID>(&self, _id: ID) -> QueryResult<BlockQueryData<Types>>
+
+    async fn get_block<ID>(&self, id: ID) -> QueryResult<BlockQueryData<Types>>
     where
         ID: Into<BlockId<Types>> + Send + Sync,
     {
-        todo!()
+        let (where_clause, param): (&str, Box<dyn ToSql + Send + Sync>) = match id.into() {
+            ResourceId::Number(n) => ("h.height = $1", Box::new(n as i64)),
+            ResourceId::Hash(h) => ("h.hash = $1", Box::new(h.to_string())),
+        };
+        // ORDER BY h.height ASC ensures that if there are duplicate blocks, we return the first one.
```

Comment:
I know there are multiple comments in the code about this but I couldn't find an open issue: 
- https://github.com/EspressoSystems/hotshot-query-service/issues/284

Reply:
Thank you